pkg/controller/gc_controller.go

```diff
@@ -46,100 +41,63 @@ const (
-	kbClient                  client.Client
-	frequency                 time.Duration
-
-	clock clock.Clock
+// gcReconciler creates DeleteBackupRequests for expired backups.
+type gcReconciler struct {
+	client.Client
+	logger logrus.FieldLogger
+	clock  clock.Clock
 }
 
-// NewGCController constructs a new gcController.
-func NewGCController(
+// NewGCReconciler constructs a new gcReconciler.
+func NewGCReconciler(
 	logger logrus.FieldLogger,
-	backupInformer velerov1informers.BackupInformer,
-	deleteBackupRequestLister velerov1listers.DeleteBackupRequestLister,
-	deleteBackupRequestClient velerov1client.DeleteBackupRequestsGetter,
-	kbClient client.Client,
-	frequency time.Duration,
-) Interface {
-	c := &gcController{
-		genericController:         newGenericController(GarbageCollection, logger),
-		clock:                     clock.RealClock{},
-		backupLister:              backupInformer.Lister(),
-		deleteBackupRequestLister: deleteBackupRequestLister,
-		deleteBackupRequestClient: deleteBackupRequestClient,
-		kbClient:                  kbClient,
+	client client.Client,
+) *gcReconciler {
+	return &gcReconciler{
+		Client: client,
+		logger: logger,
+		clock:  clock.RealClock{},
 	}
-
-	c.syncHandler = c.processQueueItem
-	c.resyncPeriod = frequency
-	if c.resyncPeriod <= 0 {
-		c.resyncPeriod = defaultGCFrequency
-	}
-	logger.Infof("Garbage collection frequency: %s", c.resyncPeriod.String())
-	c.resyncFunc = c.enqueueAllBackups
-
-	backupInformer.Informer().AddEventHandler(
-		cache.ResourceEventHandlerFuncs{
-			AddFunc:    c.enqueue,
-			UpdateFunc: func(_, obj interface{}) { c.enqueue(obj) },
-		},
-	)
-
-	return c
 }
 
-// enqueueAllBackups lists all backups from cache and enqueues all of them so we can check each one
-// for expiration.
-func (c *gcController) enqueueAllBackups() {
-	c.logger.Debug("gcController.enqueueAllBackups")
-
-	backups, err := c.backupLister.List(labels.Everything())
-	if err != nil {
-		c.logger.WithError(errors.WithStack(err)).Error("error listing backups")
-		return
-	}
-
-	for _, backup := range backups {
-		c.enqueue(backup)
-	}
+func (c *gcReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	s := kube.NewPeriodicalEnqueueSource(c.logger, mgr.GetClient(), &velerov1api.BackupList{}, defaultGCFrequency)
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&velerov1api.Backup{}).
+		Watches(s, nil).
+		Complete(c)
 }
 
-func (c *gcController) processQueueItem(key string) error {
-	log := c.logger.WithField("backup", key)
-
-	ns, name, err := cache.SplitMetaNamespaceKey(key)
-	if err != nil {
-		return errors.Wrap(err, "error splitting queue key")
-	}
-
-	backup, err := c.backupLister.Backups(ns).Get(name)
-	if apierrors.IsNotFound(err) {
-		log.Debug("Unable to find backup")
-		return nil
-	}
-	if err != nil {
-		return errors.Wrap(err, "error getting backup")
+// +kubebuilder:rbac:groups=velero.io,resources=backups,verbs=get;list;watch;update
```

Comment:
Not sure add RBAC markes here is expected, because Backup, DeleteBackupRequest, BSL's RBAC rules are already generated in their own controllers. Please check whether this will overwrite those rules.

Reply:
I understand your concern but I think adding RBAC rules here might be better. RBAC markes here won't affect current rules(won't add any new access and remove existing accesses). The purpose of doing that is to make sure gc controller continuously gets sufficient access and run normally regardless of the potential changes of RBAC rules from other controllers.

Reply:
As long as it doesn't impact existing generated config/rbac/role.yaml content, I think keeping the markers here doesn't make any harm.